### PR TITLE
[GraphOptz] Improve ReduceMean conversion to AvgPool

### LINKED ISSUE #5137: [GraphOptz] Fold Reshape into ReduceMean
We have a network with a number of `ReduceMean`s that are prevented from being optimized into `AvgPool` (see [OptimizeReduceMean](https://github.com/pytorch/glow/blob/bab9d94140a2dc371199391f997377192cc15e47/lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp#L1742)) because the input shape does not satisfy the requirement of being NCHW.

`ReduceMean` input is actually a `Reshape` from NCHW to NC{H*W}. We would like to fold this `Reshape` into into the `ReduceMean`.

For example:

```
<1 x 2 x 4 x 6>  =>  <1 x 2 x 4 x 6>
       |                    |
    Reshape                 |
  <1 x 2 x 24>              |
       |                    |
   ReduceMean    =>     ReduceMean
   (axes=[2])          (axes=[2,3])
    <1 x 2>              <1 x 2>
```
Looking for advice on where best to place this optimization.
1. As a pre-processing step in OptimizeReduceMean?
2. OptimizeReshape?
3. Something else?

### PR DESCRIPTION
**Pull request by mcurtis-cadence**
Summary:
- Handle NHWC input dimension order

- Fold Reshapes into ReduceMean to enable AvgPool opt

  AvgPool replacement requires that ReduceMean input be 4D and that
  Axes correspond to HW (of NCHW or NHWC). In some networks a
  preceeding Reshape causes the input shape to violate these
  conditions.

  This change looks for Reshapes that can be removed and folded into
  the ReduceMean so that the AvgPool optimization can be applied.

Co-authored-by: Rishab Goel <rishabg@cadence.com>

Documentation:

Fixes #5137 

Test Plan:

### CODE REVIEW

**Comment by mcurtis-cadence** [comment]
32B_DIM_T failure (ThreadPoolExecutorTest) seems unrelated to this change.

**Comment by mcurtis-cadence** [comment]
Anyone have time to review this change?

**Comment by mcurtis-cadence** [comment]
Ping

**Comment by mcurtis-cadence** [comment]
@jfix71 Ping

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 1]
```diff
@@ -1962,56 +1962,147 @@ bool OptimizeReduceMean::run(Function *F, const CompilationContext &cctx) {
...
       // In Glow, AvgPool expects NHWC.
-      auto *TR1 = F->createTranspose(
-          RM->getName().str() + ".transposeNCHW2NHWC", in, NCHW2NHWC, "NHWC");
-      auto *AP = F->createAvgPool(RM->getName().str() + ".avgPool", TR1,
-                                  kernels, strides, pads);
+      bool needsTranspose = shapeAxes[0] == 3 && shapeAxes[1] == 2; // NCHW?
+
+      // canOptReduceMeanWithAvgPool ensures that shapeAxes satisfies one of
+      // these conditions
+      assert((needsTranspose || shapeAxes[0] == 2 && shapeAxes[1] == 1) &&
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:2074]: Is there no need to check the opposite ordering for `shapeAxes`, like you do in `canOptReduceMeanWithAvgPool`?

**Review by jfix71 — COMMENTED** [review]
Cool, added a couple small comments.

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp (hunk 1 above)
Comment (mcurtis-cadence) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:2105]: `shapeAxes` gets sorted above on line 2092 so it's not needed.

**Comment by mcurtis-cadence** [comment]
@jfix71 Ping.

**Comment by aturetsk** [comment]
@jfix71, all your remarks have been addressed. Can we land this?

**Pull request closed without merging**